File: src/main/java/iog/psg/cardano/jpi/CardanoApiImpl.java
package iog.psg.cardano.jpi;

import iog.psg.cardano.CardanoApiCodec;
import scala.Enumeration;
import scala.Some;
import scala.collection.immutable.IndexedSeq;
import scala.collection.immutable.Seq;
import scala.jdk.javaapi.CollectionConverters;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.CompletionStage;


public class CardanoApiImpl implements CardanoApi {

    private final iog.psg.cardano.CardanoApi api;
    private final HelpExecute helpExecute;

    private CardanoApiImpl() {
        helpExecute = null;
        api = null;
    }

    /**
     * CardanoApi constructor
     *
     * @param api iog.psg.cardano.CardanoApi instance
     * @param helpExecute og.psg.cardano.jpi.HelpExecute instance
     */
    public CardanoApiImpl(iog.psg.cardano.CardanoApi api, HelpExecute helpExecute) {
        this.helpExecute = helpExecute;
        this.api = api;
        Objects.requireNonNull(api, "Api cannot be null");
        Objects.requireNonNull(helpExecute, "HelpExecute cannot be null");
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.Wallet> createRestore(
            String name,
            String passphrase,
            List<String> mnemonicWordList,
            int addressPoolGap) throws CardanoApiException {
        return createRestore(name, passphrase, mnemonicWordList, null, addressPoolGap);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.Wallet> createRestore(
            String name,
            String passphrase,
            List<String> mnemonicWordList,
            List<String> mnemonicSecondFactor,
            int addressPoolGap) throws CardanoApiException {
        CardanoApiCodec.MnemonicSentence mnem = createMnemonic(mnemonicWordList);

        Optional<CardanoApiCodec.MnemonicSentence> mnemonicSecondaryFactorOpt = Optional.empty();
        if (mnemonicSecondFactor != null) {
            CardanoApiCodec.MnemonicSentence mnemonicSentence = createMnemonicSecondary(mnemonicSecondFactor);
            mnemonicSecondaryFactorOpt = Optional.of(mnemonicSentence);
        }

        return helpExecute.execute(
                api.createRestoreWallet(name, passphrase, mnem, option(mnemonicSecondaryFactorOpt), option(addressPoolGap))
        );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.Wallet> createRestoreWithKey(
            String name,
            String accountPublicKey,
            int addressPoolGap
    ) throws CardanoApiException {
        return helpExecute.execute(
                api.createRestoreWalletWithKey(name, accountPublicKey, option(addressPoolGap))
        );
    }

    /**
     * {@inheritDoc}
     */
   @Override
   public CompletionStage<CardanoApiCodec.CreateTransactionResponse> createTransaction(
           String fromWalletId,
           String passphrase,
           List<CardanoApiCodec.Payment> payments,
           CardanoApiCodec.TxMetadataIn metadata,
           String withdrawal
   ) throws CardanoApiException {

        return helpExecute.execute(api.createTransaction(fromWalletId, passphrase,
                new CardanoApiCodec.Payments(CollectionConverters.asScala(payments).toSeq()),
                option(metadata),
                option(withdrawal)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.CreateTransactionResponse> createTransaction(
            String fromWalletId,
            String passphrase,
            List<CardanoApiCodec.Payment> payments
    ) throws CardanoApiException {

        return createTransaction(fromWalletId, passphrase, payments, null, "self");
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.Wallet> getWallet(
            String fromWalletId) throws CardanoApiException {

        return helpExecute.execute(
                api.getWallet(fromWalletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> deleteWallet(
            String fromWalletId) throws CardanoApiException {

        return helpExecute.execute(
                api.deleteWallet(fromWalletId)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.CreateTransactionResponse> getTransaction(
            String walletId, String transactionId) throws CardanoApiException {

        return helpExecute.execute(
                api.getTransaction(walletId, transactionId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> deleteTransaction(String walletId, String transactionId) throws CardanoApiException {
        return helpExecute.execute(api.deleteTransaction(walletId, transactionId)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.EstimateFeeResponse> estimateFee(
            String walletId, List<CardanoApiCodec.Payment> payments) throws CardanoApiException {
        return estimateFee(walletId, payments, "self", null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.EstimateFeeResponse> estimateFee(
            String walletId,
            List<CardanoApiCodec.Payment> payments,
            String withdrawal,
            CardanoApiCodec.TxMetadataIn metadata) throws CardanoApiException {

        return helpExecute.execute(
                api.estimateFee(walletId,
                        new CardanoApiCodec.Payments(CollectionConverters.asScala(payments).toSeq()),
                        option(withdrawal), option(metadata)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.FundPaymentsResponse> fundPayments(
            String walletId, List<CardanoApiCodec.Payment> payments) throws CardanoApiException {
        return helpExecute.execute(
                api.fundPayments(walletId,
                        new CardanoApiCodec.Payments(CollectionConverters.asScala(payments).toSeq())));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.WalletAddressId>> listAddresses(
            String walletId, AddressFilter addressFilter) throws CardanoApiException {

        Optional<Enumeration.Value> addressFilterOpt = Optional.empty();
        if (addressFilter != null) {
            Enumeration.Value v = CardanoApiCodec.AddressFilter$.MODULE$.Value(addressFilter.name().toLowerCase());
            addressFilterOpt = Optional.of(v);
        }

        return helpExecute.execute(
                api.listAddresses(walletId, option(addressFilterOpt))).thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.WalletAddressId>> listAddresses(
            String walletId) throws CardanoApiException {
        return listAddresses(walletId, null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.WalletAddress> inspectAddress(
            String addressId) throws CardanoApiException {
        return helpExecute.execute(api.inspectAddress(addressId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.CreateTransactionResponse>> listTransactions(
            ListTransactionsParamBuilder builder) throws CardanoApiException {
        return helpExecute.execute(
                api.listTransactions(
                        builder.getWalletId(),
                        option(builder.getStartTime()),
                        option(builder.getEndTime()),
                        builder.getOrder(),
                        option(builder.getMinwithdrawal())))
                .thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.Wallet>> listWallets() throws CardanoApiException {
        return helpExecute.execute(
                api.listWallets())
                .thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> updatePassphrase(
            String walletId,
            String oldPassphrase,
            String newPassphrase) throws CardanoApiException {

        return helpExecute.execute(api.updatePassphrase(walletId, oldPassphrase, newPassphrase)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.Wallet> updateName(
            String walletId,
            String name) throws CardanoApiException {
        return helpExecute.execute(api.updateName(walletId, name));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.NetworkInfo> networkInfo() throws CardanoApiException {
        return helpExecute.execute(api.networkInfo());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.NetworkClock> networkClock() throws CardanoApiException {
        return networkClock(null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.NetworkClock> networkClock(Boolean forceNtpCheck) throws CardanoApiException {
        return helpExecute.execute(api.networkClock(option(forceNtpCheck)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.NetworkParameters> networkParameters() throws CardanoApiException {
        return helpExecute.execute(api.networkParameters());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.UTxOStatistics> getUTxOsStatistics(String walletId) throws CardanoApiException {
        return helpExecute.execute(api.getUTxOsStatistics(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.PostExternalTransactionResponse> postExternalTransaction(String binary) throws CardanoApiException {
        return helpExecute.execute(api.postExternalTransaction(binary));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.MigrationResponse>> migrateShelleyWallet(String walletId, String passphrase, List<String> addresses) throws CardanoApiException {
        IndexedSeq<String> addressesList = CollectionConverters.asScala(addresses).toIndexedSeq();
        CompletionStage<Seq<CardanoApiCodec.MigrationResponse>> response = helpExecute.execute(api.migrateShelleyWallet(walletId, passphrase, addressesList));
        return response.thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.MigrationCostResponse> getShelleyWalletMigrationInfo(String walletId) throws CardanoApiException {
        return helpExecute.execute(api.getShelleyWalletMigrationInfo(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<CardanoApiCodec.StakePool>> listStakePools(Integer stake) throws CardanoApiException {
        CompletionStage<Seq<CardanoApiCodec.StakePool>> stakePools = helpExecute.execute(api.listStakePools(stake));
        return stakePools.thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.EstimateFeeResponse> estimateFeeStakePool(String walletId) throws CardanoApiException {
        return helpExecute.execute(api.estimateFeeStakePool(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.MigrationResponse> joinStakePool(String walletId, String stakePoolId, String passphrase) throws CardanoApiException {
        return helpExecute.execute(api.joinStakePool(walletId, stakePoolId, passphrase));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.MigrationResponse> quitStakePool(String walletId, String passphrase) throws CardanoApiException {
        return helpExecute.execute(api.quitStakePool(walletId, passphrase));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<CardanoApiCodec.StakePoolMaintenanceActionsStatus> getMaintenanceActions() throws CardanoApiException {
        return helpExecute.execute(api.getMaintenanceActions());
    }

    @Override
    public CompletionStage<Void> postMaintenanceAction() throws CardanoApiException {
        return helpExecute.execute(api.postMaintenanceAction()).thenApply(x -> null);
    }

    private static <T> scala.Option<T> option(final T value) {
        return (value != null) ? new Some<T>(value) : scala.Option.apply((T) null);
    }

    private static <T> scala.Option<T> option(final Optional<T> value) {
        return value.map(CardanoApiImpl::option).orElse(scala.Option.apply((T) null));
    }

    private static CardanoApiCodec.GenericMnemonicSentence createMnemonic(List<String> wordList) {
        return new CardanoApiCodec.GenericMnemonicSentence(
                CollectionConverters.asScala(wordList).toIndexedSeq()
        );
    }

    private static CardanoApiCodec.GenericMnemonicSecondaryFactor createMnemonicSecondary(List<String> wordList) {
        return new CardanoApiCodec.GenericMnemonicSecondaryFactor(
                CollectionConverters.asScala(wordList).toIndexedSeq()
        );
    }

}

File: src/main/scala/iog/psg/cardano/CardanoApiImpl.scala
package iog.psg.cardano

import akka.actor.ActorSystem
import akka.http.scaladsl.marshalling.{Marshal, Marshaller}
import akka.http.scaladsl.model.HttpMethods._
import akka.http.scaladsl.model.Uri.Query
import akka.http.scaladsl.model._
import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
import io.circe.generic.auto._
import io.circe.generic.extras.Configuration
import iog.psg.cardano.CardanoApi.Order.Order

import java.time.ZonedDateTime
import scala.concurrent.{ExecutionContext, Future}

private class CardanoApiImpl(baseUriWithPort: String)(implicit ec: ExecutionContext, as: ActorSystem)
    extends CardanoApi {

  import CardanoApiCodec.ImplicitCodecs._
  import CardanoApiCodec._
  import AddressFilter.AddressFilter
  import iog.psg.cardano.CardanoApi._

  private val addresses = s"${baseUriWithPort}addresses"
  private val proxy = s"${baseUriWithPort}proxy"
  private val wallets = s"${baseUriWithPort}wallets"
  private val network = s"${baseUriWithPort}network"
  private val stakePools = s"${baseUriWithPort}stake-pools"
  private def generateMigrationsUrl(walletId: String) = s"$wallets/$walletId/migrations"

  implicit val config: Configuration = Configuration.default.withSnakeCaseMemberNames

  /**
   * @inheritdoc
   */
  override def listWallets: CardanoApiRequest[Seq[Wallet]] = CardanoApiRequest(
    HttpRequest(
      uri = wallets,
      method = GET
    ),
    _.toWallets
  )

  /**
   * @inheritdoc
   */
  override def getWallet(walletId: String): CardanoApiRequest[Wallet] = CardanoApiRequest(
    HttpRequest(
      uri = s"$wallets/$walletId",
      method = GET
    ),
    _.toWallet
  )

  /**
   * @inheritdoc
   */
  override def updateName(walletId: String, name: String): Future[CardanoApiRequest[Wallet]] = {
    val body = Map("name" -> name)
    Marshal(body).to[RequestEntity].map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$wallets/$walletId",
          entity = marshalled,
          method = PUT
        ),
        _.toWallet
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def networkInfo: CardanoApiRequest[NetworkInfo] = CardanoApiRequest(
    HttpRequest(
      uri = s"${network}/information",
      method = GET
    ),
    _.toNetworkInfoResponse
  )

  /**
   * @inheritdoc
   */
  override def networkClock(forceNtpCheck: Option[Boolean] = None): CardanoApiRequest[NetworkClock] = {
    val url = s"$network/clock${forceNtpCheck.map(force => s"?forceNtpCheck=$force").getOrElse("")}"
    CardanoApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toNetworkClockResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def networkParameters(): CardanoApiRequest[NetworkParameters] =
    CardanoApiRequest(
      HttpRequest(
        uri = s"${network}/parameters",
        method = GET
      ),
      _.toNetworkParametersResponse
    )

  private def prepareCreateRestoreRequest[T](
    body: T
  )(implicit m: Marshaller[T, RequestEntity]) =
    Marshal(body).to[RequestEntity].map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$wallets",
          method = POST,
          entity = marshalled
        ),
        _.toWallet
      )
    }

  /**
   * @inheritdoc
   */
  override def createRestoreWallet(name: String,
                                   passphrase: String,
                                   mnemonicSentence: MnemonicSentence,
                                   mnemonicSecondFactor: Option[MnemonicSentence] = None,
                                   addressPoolGap: Option[Int] = None
  ): Future[CardanoApiRequest[Wallet]] =
    prepareCreateRestoreRequest(
      CreateRestore(
        name,
        passphrase,
        mnemonicSentence.mnemonicSentence,
        mnemonicSecondFactor.map(_.mnemonicSentence),
        addressPoolGap
      )
    )

  /**
   * @inheritdoc
   */
  override def createRestoreWalletWithKey(name: String,
                                          accountPublicKey: String,
                                          addressPoolGap: Option[Int] = None
  ): Future[CardanoApiRequest[Wallet]] =
    prepareCreateRestoreRequest(
      CreateRestoreWithKey(
        name,
        accountPublicKey,
        addressPoolGap
      )
    )

  /**
   * @inheritdoc
   */
  override def listAddresses(walletId: String,
                             state: Option[AddressFilter]
  ): CardanoApiRequest[Seq[WalletAddressId]] = {

    val baseUri = Uri(s"$wallets/${walletId}/addresses")

    val url = state.map { s =>
      baseUri.withQuery(Query("state" -> s.toString))
    }.getOrElse(baseUri)

    CardanoApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toWalletAddressIds
    )

  }

  /**
   * @inheritdoc
   */
  override def inspectAddress(addressId: String): CardanoApiRequest[WalletAddress] = {
    val url = Uri(s"$addresses/$addressId")

    CardanoApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toWalletAddress
    )
  }

  /**
   * @inheritdoc
   */
  override def listTransactions(walletId: String,
                                start: Option[ZonedDateTime] = None,
                                end: Option[ZonedDateTime] = None,
                                order: Order = Order.descendingOrder,
                                minWithdrawal: Option[Int] = None
  ): CardanoApiRequest[Seq[CreateTransactionResponse]] = {
    val baseUri = Uri(s"$wallets/${walletId}/transactions")

    val queries =
      Seq("start", "end", "order", "minWithdrawal").zip(Seq(start, end, order, minWithdrawal)).collect {
        case (queryParamName, order: Order)            => queryParamName -> order.toString
        case (queryParamName, Some(dt: ZonedDateTime)) => queryParamName -> zonedDateToString(dt)
        case (queryParamName, Some(minWith: Int))      => queryParamName -> minWith.toString
      }

    val uriWithQueries = baseUri.withQuery(Query(queries: _*))

    CardanoApiRequest(
      HttpRequest(
        uri = uriWithQueries,
        method = GET
      ),
      _.toCreateTransactionsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def createTransaction(fromWalletId: String,
                                 passphrase: String,
                                 payments: Payments,
                                 metadata: Option[TxMetadataIn],
                                 withdrawal: Option[String]
  ): Future[CardanoApiRequest[CreateTransactionResponse]] = {

    val createTx = CreateTransaction(passphrase, payments.payments, metadata, withdrawal)

    Marshal(createTx).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$wallets/$fromWalletId/transactions",
          method = POST,
          entity = marshalled
        ),
        _.toCreateTransactionResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def estimateFee(fromWalletId: String,
                           payments: Payments,
                           withdrawal: Option[String],
                           metadataIn: Option[TxMetadataIn] = None
  ): Future[CardanoApiRequest[EstimateFeeResponse]] = {

    val estimateFees = EstimateFee(payments.payments, withdrawal, metadataIn)

    Marshal(estimateFees).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$wallets/$fromWalletId/payment-fees",
          method = POST,
          entity = marshalled
        ),
        _.toEstimateFeeResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def fundPayments(walletId: String,
                            payments: Payments
  ): Future[CardanoApiRequest[FundPaymentsResponse]] =
    Marshal(payments).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$wallets/${walletId}/coin-selections/random",
          method = POST,
          entity = marshalled
        ),
        _.toFundPaymentsResponse
      )
    }

  /**
   * @inheritdoc
   */
  override def getTransaction[T <: TxMetadataIn](walletId: String,
                                                 transactionId: String
  ): CardanoApiRequest[CreateTransactionResponse] = {

    val uri = Uri(s"$wallets/${walletId}/transactions/${transactionId}")

    CardanoApiRequest(
      HttpRequest(
        uri = uri,
        method = GET
      ),
      _.toCreateTransactionResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def deleteTransaction(walletId: String, transactionId: String): CardanoApiRequest[Unit] = {
    val uri = Uri(s"$wallets/${walletId}/transactions/${transactionId}")

    CardanoApiRequest(
      HttpRequest(
        uri = uri,
        method = DELETE
      ),
      _.toUnit
    )
  }

  /**
   * @inheritdoc
   */
  override def updatePassphrase(walletId: String,
                                oldPassphrase: String,
                                newPassphrase: String
  ): Future[CardanoApiRequest[Unit]] = {

    val uri = Uri(s"$wallets/${walletId}/passphrase")
    val updater = UpdatePassphrase(oldPassphrase, newPassphrase)

    Marshal(updater).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = uri,
          method = PUT,
          entity = marshalled
        ),
        _.toUnit
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def deleteWallet(
    walletId: String
  ): CardanoApiRequest[Unit] = {

    val uri = Uri(s"$wallets/${walletId}")

    CardanoApiRequest(
      HttpRequest(
        uri = uri,
        method = DELETE
      ),
      _.toUnit
    )
  }

  /**
   * @inheritdoc
   */
  override def getUTxOsStatistics(walletId: String): CardanoApiRequest[UTxOStatistics] = {
    val uri = Uri(s"$wallets/$walletId/statistics/utxos")
    CardanoApiRequest(
      HttpRequest(
        uri = uri,
        method = GET
      ),
      _.toUTxOStatisticsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def postExternalTransaction(binary: String): CardanoApiRequest[PostExternalTransactionResponse] = {
    val uri = Uri(s"$proxy/transactions")
    CardanoApiRequest(
      HttpRequest(
        uri = uri,
        method = POST,
        entity = HttpEntity(binary).withContentType(ContentTypes.`application/octet-stream`)
      ),
      _.toPostExternalTransactionResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def migrateShelleyWallet(walletId: String,
                                    passphrase: String,
                                    addresses: Seq[String]
  ): Future[CardanoApiRequest[Seq[MigrationResponse]]] = {
    val updater = SubmitMigration(passphrase = passphrase, addresses = addresses)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = generateMigrationsUrl(walletId),
          method = POST,
          entity = marshalled
        ),
        _.toSubmitMigrationsResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def getShelleyWalletMigrationInfo(walletId: String): CardanoApiRequest[MigrationCostResponse] =
    CardanoApiRequest(
      HttpRequest(
        uri = generateMigrationsUrl(walletId),
        method = GET
      ),
      _.toMigrationCostResponse
    )

  /**
   * @inheritdoc
   */
  override def listStakePools(stake: Int): CardanoApiRequest[Seq[StakePool]] = {
    val url = Uri(s"$stakePools?stake=$stake")
    CardanoApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toStakePoolsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def estimateFeeStakePool(walletId: String): CardanoApiRequest[EstimateFeeResponse] = {
    val url = Uri(s"$wallets/$walletId/delegation-fees")
    CardanoApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toEstimateFeeResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def joinStakePool(walletId: String,
                             stakePoolId: String,
                             passphrase: String
  ): Future[CardanoApiRequest[MigrationResponse]] = {
    val updater = PassphraseRequest(passphrase = passphrase)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$stakePools/$stakePoolId/wallets/$walletId",
          method = PUT,
          entity = marshalled
        ),
        _.toSubmitMigrationResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def quitStakePool(walletId: String,
                             passphrase: String
  ): Future[CardanoApiRequest[MigrationResponse]] = {
    val updater = PassphraseRequest(passphrase = passphrase)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$stakePools/*/wallets/$walletId",
          method = DELETE,
          entity = marshalled
        ),
        _.toSubmitMigrationResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def getMaintenanceActions(): CardanoApiRequest[StakePoolMaintenanceActionsStatus] =
    CardanoApiRequest(
      HttpRequest(
        uri = s"$stakePools/maintenance-actions",
        method = GET
      ),
      _.toStakePoolMaintenanceActionsStatusResponse
    )

  /**
   * @inheritdoc
   */
  override def postMaintenanceAction(): Future[CardanoApiRequest[Unit]] = {
    val action = PostMaintenanceActionRequest("gc_stake_pools")
    Marshal(action).to[RequestEntity] map { marshalled =>
      CardanoApiRequest(
        HttpRequest(
          uri = s"$stakePools/maintenance-actions",
          method = POST,
          entity = marshalled
        ),
        _.toUnit
      )
    }
  }

}
